### ISSUE TITLE
[partially defined] Doesn't catch usage within an if statement

### ISSUE THREAD

**Issue by hauntsaninja** [python/mypy#14124]
I'd expect an error in the following when using `--enable-error-code partially-defined`:
```
def bar(x: int) -> None: ...

def foo(cond: bool) -> None:
    if cond:
        a = 5
    else:
        bar(a)
    # print(a + 1)
```
Note that we get an error on the print statement if we uncomment it.

cc @ilinum

**Comment by ilinum** [comment]
I think this is part of general type of errors of the form that are currently not detected by mypy at all:
```python
def foo():
  x = y
  y = 1
```

@ilevkivskyi suggested recently that we could modify the partially-defined check to detect variables that are used before definition. I believe this should be possible.

Let's use this issue to track this request.
